src/core/tsi/alts/handshaker/alts_handshaker_client.cc

```diff
@@ -358,6 +335,164 @@ static tsi_result make_grpc_call(alts_handshaker_client* c, bool is_start) {
+      head_ = nullptr;
+    }
+  }
+
+  grpc_core::Combiner* combiner_ = grpc_combiner_create();
+  Node* head_ = nullptr;
+  Node* tail_ = nullptr;
+  size_t outstanding_handshakes_ = 0;
+  const size_t max_outstanding_handshakes_;
+};
+
+gpr_once g_queued_handshakes_init = GPR_ONCE_INIT;
+/* Using separate queues for client and server handshakes is a
+ * hack that's mainly intended to satisfy the alts_concurrent_connectivity_test,
+ * which runs many concurrent handshakes where both endpoints
+ * are in the same process; this situation is problematic with a
+ * single queue because we have a high chance of using up all outstanding
+ * slots in the queue, such that there aren't any
+ * mutual client/server handshakes outstanding at the same time and
+ * able to make progress. */
+HandshakeQueue* g_client_handshake_queue;
+HandshakeQueue* g_server_handshake_queue;
+
+void DoHandshakeQueuesInit(void) {
+  g_client_handshake_queue =
+      grpc_core::New<HandshakeQueue>(40 /* max outstanding handshakes */);
+  g_server_handshake_queue =
+      grpc_core::New<HandshakeQueue>(40 /* max outstanding handshakes */);
+}
+
+void EnqueueHandshakeAndMaybeStartSome(alts_grpc_handshaker_client* client,
+                                       bool is_client) {
+  gpr_once_init(&g_queued_handshakes_init, DoHandshakeQueuesInit);
+  if (is_client) {
+    g_client_handshake_queue->EnqueueHandshakeAndMaybeStartSome(client);
+  } else {
+    g_server_handshake_queue->EnqueueHandshakeAndMaybeStartSome(client);
+  }
+}
+
+};  // namespace
+
+/**
+ * Populate grpc operation data with the fields of ALTS handshaker client and
+ * make a grpc call.
+ */
+static tsi_result make_grpc_call(alts_handshaker_client* c, bool is_start) {
+  GPR_ASSERT(c != nullptr);
+  alts_grpc_handshaker_client* client =
+      reinterpret_cast<alts_grpc_handshaker_client*>(c);
+  if (is_start) {
+    EnqueueHandshakeAndMaybeStartSome(client, client->is_client);
+    return TSI_OK;
+  } else {
+    return continue_make_grpc_call(client, is_start);
+  }
+}
+
+static void on_status_received(void* arg, grpc_error* error) {
+  alts_grpc_handshaker_client* client =
+      static_cast<alts_grpc_handshaker_client*>(arg);
+  if (client->handshake_status_code != GRPC_STATUS_OK) {
+    // TODO(apolcyn): consider overriding the handshake result's
+    // status from the final ALTS message with the status here.
+    char* status_details =
+        grpc_slice_to_c_string(client->handshake_status_details);
+    gpr_log(GPR_INFO,
+            "alts_grpc_handshaker_client:%p on_status_received "
+            "status:%d details:|%s| error:|%s|",
+            client, client->handshake_status_code, status_details,
+            grpc_error_string(error));
+    gpr_free(status_details);
+  }
+  maybe_complete_tsi_next(client, true /* receive_status_finished */,
+                          nullptr /* pending_recv_message_result */);
+  EnqueueHandshakeAndMaybeStartSome(nullptr, client->is_client);
```

Comment:
I believe this serves to free up one slot in the outstanding queue?

Reply:
Correct, added a comment to this function to explain.